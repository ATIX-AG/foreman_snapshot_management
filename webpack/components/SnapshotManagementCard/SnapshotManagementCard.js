import PropTypes from 'prop-types';
import React from 'react';

import CardTemplate from 'foremanReact/components/HostDetails/Templates/CardItem/CardTemplate';
import { translate as __ } from 'foremanReact/common/I18n';

import SnapshotManagement from '../SnapshotManagement';

const SnapshotManagementCard = ({ hostDetails, ...props }) => {
  const children = [];
  const capabilities = {
    editableSnapshotName:
      hostDetails?.capabilities?.includes('editable_snapshot_name') || false,
    limitSnapshotNameFormat:
      hostDetails?.capabilities?.includes('limit_snapshot_name_format') ||
      false,
    quiesceOption:
      hostDetails?.capabilities?.includes('snapshot_include_quiesce') || false,
  };

  if (hostDetails?.uuid && hostDetails?.id && hostDetails?.permissions) {
    children.push(
      <SnapshotManagement
        key="SnapshotManagement"
        host={{ id: hostDetails.id, name: hostDetails.name }}
        canCreate={hostDetails.permissions.create_snapshots}
        canUpdate={hostDetails.permissions.edit_snapshots}
        canRevert={hostDetails.permissions.revert_snapshots}
        canDelete={hostDetails.permissions.destroy_snapshots}
        capabilities={capabilities}
      />
    );
    return (
      <CardTemplate
        overrideGridProps={{ xl2: 6, xl: 8, lg: 8, md: 12 }}
        header={__('Snapshots')}
      >
        {children}
      </CardTemplate>
    );
  }
  return null;
};

export default SnapshotManagementCard;

SnapshotManagementCard.propTypes = {
  hostDetails: PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
    uuid: PropTypes.string,
    capabilities: PropTypes.array,
    permissions: PropTypes.shape({
      create_snapshots: PropTypes.bool,
      edit_snapshots: PropTypes.bool,
      revert_snapshots: PropTypes.bool,
      destroy_snapshots: PropTypes.bool,
    }),
  }),
};

SnapshotManagementCard.defaultProps = {
  hostDetails: {},
};
