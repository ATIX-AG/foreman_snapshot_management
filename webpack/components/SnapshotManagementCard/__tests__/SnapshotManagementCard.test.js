import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import SnapshotManagementCard from '../SnapshotManagementCard';

const hostDetails = {
  id: 42,
  name: 'deep.thought',
  uuid: '829aa26a-928f-11ee-b9d1-0242ac120002',
  permissions: {
    create_snapshots: true,
    edit_snapshots: false,
    revert_snapshots: true,
    destroy_snapshots: false,
  },
};

const hostDetailsVmware = {
  ...hostDetails,
  capabilities: [
    'build',
    'image',
    'snapshots',
    'snapshot_include_quiesce',
    'snapshot_include_ram',
    'editable_snapshot_name',
  ],
};
const hostDetailsProxmox = {
  ...hostDetails,
  capabilities: [
    'build',
    'new_volume',
    'new_interface',
    'image',
    'snapshots',
    'limit_snapshot_name_format',
  ],
};

const fixtures = {
  'without optional Props': {},
  'with Props': { hostDetails },
  'with VMWare capabilities': {
    hostDetails: hostDetailsVmware,
  },
  'with Proxmox capabilities': {
    hostDetails: hostDetailsProxmox,
  },
};

describe('SnapshotManagementCard', () => {
  describe('renders', () =>
    testComponentSnapshotsWithFixtures(SnapshotManagementCard, fixtures));
});
