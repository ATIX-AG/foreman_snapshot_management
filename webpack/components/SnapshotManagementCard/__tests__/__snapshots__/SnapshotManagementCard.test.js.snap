// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapshotManagementCard renders with NotSupported CR capabilities 1`] = `""`;

exports[`SnapshotManagementCard renders with Props 1`] = `
<CardTemplate
  expandable={false}
  header="Snapshots"
  masonryLayout={false}
  overrideDropdownProps={Object {}}
  overrideGridProps={
    Object {
      "lg": 8,
      "md": 12,
      "xl": 8,
      "xl2": 6,
    }
  }
>
  <Connect(SnapshotManagement)
    canCreate={true}
    canDelete={false}
    canRevert={true}
    canUpdate={false}
    capabilities={
      Object {
        "editableSnapshotName": false,
        "limitSnapshotNameFormat": false,
        "quiesceOption": false,
      }
    }
    host={
      Object {
        "id": 42,
        "name": "deep.thought",
      }
    }
    key="SnapshotManagement"
  />
</CardTemplate>
`;

exports[`SnapshotManagementCard renders with Proxmox capabilities 1`] = `
<CardTemplate
  expandable={false}
  header="Snapshots"
  masonryLayout={false}
  overrideDropdownProps={Object {}}
  overrideGridProps={
    Object {
      "lg": 8,
      "md": 12,
      "xl": 8,
      "xl2": 6,
    }
  }
>
  <Connect(SnapshotManagement)
    canCreate={true}
    canDelete={false}
    canRevert={true}
    canUpdate={false}
    capabilities={
      Object {
        "editableSnapshotName": false,
        "limitSnapshotNameFormat": true,
        "quiesceOption": false,
      }
    }
    host={
      Object {
        "id": 42,
        "name": "deep.thought",
      }
    }
    key="SnapshotManagement"
  />
</CardTemplate>
`;

exports[`SnapshotManagementCard renders with VMWare capabilities 1`] = `
<CardTemplate
  expandable={false}
  header="Snapshots"
  masonryLayout={false}
  overrideDropdownProps={Object {}}
  overrideGridProps={
    Object {
      "lg": 8,
      "md": 12,
      "xl": 8,
      "xl2": 6,
    }
  }
>
  <Connect(SnapshotManagement)
    canCreate={true}
    canDelete={false}
    canRevert={true}
    canUpdate={false}
    capabilities={
      Object {
        "editableSnapshotName": true,
        "limitSnapshotNameFormat": false,
        "quiesceOption": true,
      }
    }
    host={
      Object {
        "id": 42,
        "name": "deep.thought",
      }
    }
    key="SnapshotManagement"
  />
</CardTemplate>
`;

exports[`SnapshotManagementCard renders without optional Props 1`] = `""`;
